src/net/peer/connect/bootstrap/bootstrap.rs

```diff
@@ -64,7 +64,7 @@ pub fn bootstrap<UID: Uid>(
     ext_reachability: ExternalReachability,
     blacklist: HashSet<SocketAddr>,
     use_service_discovery: bool,
-    config: ConfigFile,
+    config: &ConfigFile,
```

Comment:
When changing everything to reference, also keep in mind that "Copy" types are better passed by value (and hence copied) - at-least if it's not huge arrays inside structs etc. Most of the time they will be NRVO optimised out unlike references which make it relatively more opaque to optimisations depending on scenarios. Don't know if it applies to rust, but certainly was something i remember from C/C++ days (and i have forgotten most of it anyway). Refer to literature like [these](https://stackoverflow.com/a/40185996/1060004) and NRVOs etc.

Could you pls scan and confirm this and correct it if necessary ?

Reply:
Well, this seems like a tricky question.
I couldn't find a definite answer regarding NVO in Rust. 
Some people claim it's there (LLVM might be doing that): https://stackoverflow.com/a/27837182/3323667.
I also seems like some related work is still in progress: https://github.com/rust-lang/rust/issues/32966.

I was following `clippy` lint https://rust-lang-nursery.github.io/rust-clippy/v0.0.166/index.html#needless_pass_by_value:
```
warning: this argument is passed by value, but not consumed in the function body
  --> src/net/peer/connect/bootstrap/bootstrap.rs:67:13
   |
67 |     config: ConfigFile,
   |             ^^^^^^^^^^ help: consider taking a reference instead: `&ConfigFile`
   |
   = note: #[warn(needless_pass_by_value)] on by default
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.166/index.html#needless_pass_by_value
```

And I still don't know how using reference might be worse than consuming a value.

Reply:
It's worse because of the reasons stated. Pointers are bad for stuffs like cache-locality etc. Also if taking something as `var: const i32*` will certainly lead to more code generation than `var: i32` due to load semantics. Since references are pointers behind the scene, same applies to them.

The reason why clippy complains here is because [ConfigFile is non-copy type](https://github.com/maidsafe/crust/blob/e77e71cf7f96e60c41c5403d72558f52d68de8e6/src/config.rs#L32), and for those things you can only move, clone or pass reference. In this case it makes sense to pass reference. That is why the review comment explicitly mentioned
> ... "Copy" types are better passed by value ...

Anyway do keep a note to make sure that we are not always using references but checking if it's a Copy type and then evaluating it what should we use (which was the point of the review comment - to scan and check we are just not rampantly doing it for everything).

Reply:
yeah, it's `tests::compat_api::start_two_services_exchange_data` which hangs.
It already does in `dev` branch, so I'd like to research and fix it with another PR.

Reply:
also, `ConfigFile` itself is `Arc`
```
/// A handle to a crust config file. This handle can be cloned and shared throughout the program.
#[derive(Clone)]
pub struct ConfigFile {
    inner: Arc<RwLock<ConfigWrapper>>,
}
```
so it's definitely no point of referencing to it.

Reply:
Referencing is better than cloning - cloning will unnecessary cause an instruction emission for atomic increase of reference count, which in this case, being and Arc instead of Rc, will be atomic (which is certainly not a trivial as for Rc). Then the local destruction will cause another one to reduce the atomic reference count, which is again not optimal. Either or both of these would be sequentially consistent and thus interfere with optimisers. Anything that involves thread safety is not really optimal and hence should be used only if thread safety is required, else it's like having a mutex locked and unlocked in a single threaded system - buys nothing but wastes cpu cycles and prevents optimisations (they are like memory (re-ordering) barriers).

Hence (just my opinion) Arc's should possible be passed by reference unless sending over threads where ofc you should pass a clone (and safe Rust will not allow you to do otherwise anyway).

Reply:
that makes sense :)